drivers/gpu/drm/vc4/vc4_vec.c

```diff
@@ -645,6 +645,9 @@ static void vc4_vec_encoder_enable(struct drm_encoder *encoder)
 		return;
 	}
 
+	eff_clk_rate = clk_get_rate(vec->clock);
+	DRM_DEBUG_DRIVER("Effective clock rate: %ld\n", eff_clk_rate);
```

Comment:
@popcornmix Do you believe we'll have this level of control over the VEC clock? I thought we locked it at 108MHz for the VEC having ensured we were doing an integer divide to avoid jitter.

It's clock ```clocks = <&clocks BCM2835_CLOCK_VEC>```, so it is the bcm2835-cprman driver and not the firmware clocks.

Reply:
The vec will always be 108MHz following https://github.com/raspberrypi/firmware/commit/2984c5372e3fcaf2ca1b2c4b08eedb94a3a888f44051db861ab59391c77d170a
It used to be that the vec clock shared a PLL with core clock, which was affected by `enable_tvout` or `hdmi_enable_4kp60`, and the 108MHz was only accurate with `enable_tvout`.
Now it uses a dedicated PLL and will always be 108MHz, so I feel handling it not being 108MHz is redundant.

Reply:
Sorry, I quoted a line which was reading back the clock having set it to ```8000 * adjusted_mode->clock``` a few lines earlier, thereby enabling support for pixel clocks other than 13.5MHz.

You say it's on a dedicated PLL, but is that shared with anything else? (Your link to the firmware change isn't working for me to check myself).

Reply:
I've corrected the firmware commit. VEC clock will always be 108MHz.
It comes from PLLC which is fixed at 2592MHz (so always a multiple of 108MHz).
It's also used by CLK108 (which unsurprisingly also likes 108MHz) used by HDMI.

Reply:
OK, thanks. I think that makes the last commit ineffective then.
AIUI the VEC will produce the pixel clock to drive the PV, which drives HVS, so if that only ever gets 108MHz then the pixel clock is fixed.

It is talking to bcm2835-cprman, which believes it knows how to configure the VEC clock. There's a comment of ```Only operating frequency is 108Mhz``` and PLLH_AUX.
That commit is for 2711 only. Is it feasible on earlier chips?

Reply:
You should get exactly 108MHz for the vec on any Pi and with any config.txt settings with recent firmware.
If you've run `rpi-update` since 24th March you will have this behaviour.

Reply:
First of all, sorry for going silent on this for so long.

Anyway, I rebased this PR and started fresh with testing:
* I flashed freshly downloaded `2021-05-07-raspios-buster-armhf-full` image onto an SD card
* Updated everything via `apt`
* Ran `rpi-update` - `/boot/.firmware_revision` reads `bad5282375a2d1e4947432bd80a8839fbc3504ae` now
* Recompiled the kernel from the head of this PR as pushed just now and installed it

...and the results are the same as always. On Pi3, I can manipulate the pixel clock with quite an impressive precision; on Pi4, it is much more coarse, but I still have quite some control (that "405-line" mode from the example in the original post is more like 49 Hz instead of 50 Hz on Pi4, according to my oscilloscope). Without `enable_tvout=1` in config, the default modes are still about 0.8% slow, which causes unusable colour on the default branch, and this branch makes the colours look acceptable (albeit with more noticeable cross-colour artifacts than with `enable_tvout=1`).

Reply:
Can you explain what mechanism you are using to manipulate pixel clock?
Note, the difference between Pi3 and Pi4 is the Pi3 will default to driving composite when there is no hdmi connected. Pi4 will only drive composite with `enable_sdtv=1`, so add `hdmi_force_hotplug=1` to both systems for a fair comparison.

The PLLs on Pi3 and Pi4 (with recent firmware) are both multiples of 108MHz, so you should be able to get an accurate 108Mhz clock on either if the VEC clock is configured. But it won't be configured unless someone has requested it.

The linux driver calls `clk_set_rate(vec->clock, 108000000)` but I think that is done using the (deprecated) clk-bcm2835 clock driver which is likely to be using the wrong PLL, so can't get an accurate 108MHz.  It should be switched to clk-raspberrypi, the newer firmware clock driver (used by core, pixel, arm. hvs etc) which does know which PLL to use.

Reply:
To be honest, the clocking APIs is not something I'm too well versed in. I did not touch the _flow_ of clocking in the driver in any way, I did not attempt to thoroughly understand it; I just changed that hardcoded 108 MHz to a value derived from the modeline's pixel clock, saw it working to a reasonable extent and was satisfied with that.

Note that the official kernel behaves in the same way - on Pi4 without `enable_sdtv=1`, if you enable the KMS driver with composite enabled, the clock is not quite 108 MHz - but without the modifications I made here, it means broken colours.

Trying to analyse the unfamiliar code, it looks that the clock is selected in the DTS (for [bcm2835](https://github.com/raspberrypi/linux/blob/6237d09759ba6c8746cd1e19b16faee50c97bfac/arch/arm/boot/dts/bcm2835-common.dtsi#L112) and [bcm2711](https://github.com/raspberrypi/linux/blob/6237d09759ba6c8746cd1e19b16faee50c97bfac/arch/arm/boot/dts/bcm2711.dtsi#L306)) and indeed configured in [`clk-bcm2835.c`](https://github.com/raspberrypi/linux/blob/6237d09759ba6c8746cd1e19b16faee50c97bfac/drivers/clk/bcm/clk-bcm2835.c#L2218). There doesn't seem to be any difference in configuration between the BCM2711 and BCM2835 variants.

There seems to be the [`clk-raspberrypi.c`](https://github.com/raspberrypi/linux/blob/6237d09759ba6c8746cd1e19b16faee50c97bfac/drivers/clk/bcm/clk-raspberrypi.c) file you're referencing, but to be honest I have no idea what it does beyond acknowledging the existence of firmware clocks and how to make use of it.

I will do testing with `hdmi_force_hotplug=1` probably today in the evening to see whether that affects anything.

Reply:
Something weird happens on my Pi3 with `hdmi_force_hotplug=1` - X starts fine on the composite output during the initial boot, but the driver freezes if I try to change modes using xrandr. Curiously, even attempting to set a mode that would be rejected by `vc4_vec_connector_atomic_check()` (causing `xrandr: Configure crtc 3 failed` when `hdmi_force_hotplug=0`), causes this freeze. In dmesg, there's only this:

```
[   82.394798] [drm:drm_atomic_helper_wait_for_flip_done [drm_kms_helper]] *ERROR* [CRTC:83:crtc-3] flip_done timed out
[   92.634957] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CRTC:83:crtc-3] flip_done timed out
[  102.875033] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CONNECTOR:49:Composite-1] flip_done timed out
[  113.115075] [drm:drm_atomic_helper_wait_for_flip_done [drm_kms_helper]] *ERROR* [CRTC:83:crtc-3] flip_done timed out
[  123.355119] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CRTC:83:crtc-3] flip_done timed out
[  133.595143] [drm:drm_atomic_helper_wait_for_flip_done [drm_kms_helper]] *ERROR* [CRTC:83:crtc-3] flip_done timed out
```

However, that also happens with the official kernel (both the newest one from `apt` and the newest one from `rpi-update`), so this is most likely orthogonal to what's going on in this PR.

----

Otherwise, there is not much difference regarding what we're talking about here. On Pi3, the clock is still perfectly steady (the colors look fine on the official kernel), and if I change the clock (which due to the bug above I can only do by hardcoding some other value in the module sources), it is respected precisely.

Pi4 boots up with perfect 13.5 MHz pixel clock when both `hdmi_force_hotplug=1` and `enable_tvout=1` are enabled. In fact, `hdmi_force_hotplug=1` seems to be completely ignored in this case, as I still get the rainbow test pattern on composite at boot.

However, I just discovered something curious that I overlooked previously. When I subsequently tried to change the pixel clock to 12.345 MHz, I ended up with something around 11.718 MHz instead, according to my measurements (I didn't check what the kernel reports, just measured the sync signals on output) - but that's to be expected given the coarser clock on Pi4. The curious part is that if I then try to go back to 13.5 MHz, I end up on that 13.393 or so MHz frequency - the same that is the result of `enable_tvout=0` - even though I had perfect 13.5 MHz at boot time. This is reproducible with `enable_tvout=1` regardless of whether `hdmi_force_hotplug` is enabled or not.

When `hdmi_force_hotplug=1` is set while `enable_tvout=0` on Pi4, the composite output comes up with vertical and horizontal sync patterns implying that imperfect 13.393 MHz pixel clock, but the driver freezes on a black screen and I never get to the desktop. `dmesg` shows similar problems as when attempting to switch modes on Pi3:

```
[   70.875413] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CRTC:65:crtc-1] flip_done timed out
[   81.115403] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CONNECTOR:45:Composite-1] flip_done timed out
[   91.355389] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [PLANE:59:plane-1] flip_done timed out
[  101.595385] [drm:drm_atomic_helper_wait_for_flip_done [drm_kms_helper]] *ERROR* [CRTC:65:crtc-1] flip_done timed out
[  111.835388] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CRTC:65:crtc-1] flip_done timed out
[  122.075157] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CONNECTOR:45:Composite-1] flip_done timed out
[  132.312235] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [PLANE:59:plane-1] flip_done timed out
[  142.550714] [drm:drm_atomic_helper_wait_for_flip_done [drm_kms_helper]] *ERROR* [CRTC:65:crtc-1] flip_done timed out
[  152.789930] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CRTC:65:crtc-1] flip_done timed out
```

Again, this happens even on the official kernel from `apt`, so it's not related to this diff at all.

And similar to a Pi3, if I hardcode a different clock speed in the kernel, it is respected at boot time, albeit with Pi4's normal lower accuracy (so 11.718 MHz instead of intended 12.345 MHz) - implied by sync signals on that blank screen, that is.

----

So, to sum up: we've discovered an unrelated bug, and a curious relation between `enable_tvout=1` and changing clock rates at runtime, but otherwise it still doesn't line up with that "locked 108MHz" behavior you seem to hope for.

Reply:
By the way, here's how composite output with `enable_tvout=0` looks like on the current official kernels in [NTSC](https://user-images.githubusercontent.com/4499762/136106805-5881e5fc-36ce-41e8-a583-4f154ddf70d8.png) and [PAL](https://user-images.githubusercontent.com/4499762/136106855-3dfe100b-a360-4da0-9aed-22860b38700b.png) modes, at least through my grabber. The line frequency difference is little enough that the luminance component is clear, but colour is all over the place.

With this diff in place, the chroma subcarrier frequency is compensated in this case, which makes the image look mostly right, just with some more noticeable cross-colour artifacting, especially noticeable on fine details such as text: [`enable-tvout=0`](https://user-images.githubusercontent.com/4499762/136108048-548d98bd-8606-4e5c-8587-435f81f12457.png) vs. [`enable_tvout=1`](https://user-images.githubusercontent.com/4499762/136108299-d3ca7c0d-ccb4-451d-b214-f0f244e71ad4.png) (these are NTSC; there are far less artifacts with PAL anyway). You can note the slight image width difference on these screenshots, too.